### PR TITLE
 rgw: cls_bucket_list_unordered lists a single shard

### PR DESCRIPTION
**Pull request by cbodley**
`RGWRados::cls_bucket_list_unordered()` loops over bucket index shards, but each loop uses `CLSRGWIssueBucketList` to send a `[call rgw.bucket_list]` osd op to every shard of the bucket index. this change exposes a new `cls_rgw_bucket_list()` in cls_rgw_client.h, and `RGWRados::cls_bucket_list_unordered()` uses that to list a single shard object by oid

before/after logs of an unordered list_bucket request from s3tests_boto3.functional.test_s3.test_bucket_list_unordered with rgw_override_bucket_index_max_shards=8:

- [unordered-8-shard-bucket-list-before.log](https://gist.github.com/cbodley/7783ebc0ba850c1fc679edc19d495e38)
- [unordered-8-shard-bucket-list-after.log](https://gist.github.com/cbodley/fd472e22180f331c3101f6864387dfee)

the difference is clear after counting the number of `[call rgw.bucket_list]` osd ops:
> $ grep 'call rgw.bucket_list' unordered-8-shard-bucket-list-before.log | wc -l                                                        
64
$ grep 'call rgw.bucket_list' unordered-8-shard-bucket-list-after.log | wc -l                                                        
8

Fixes: http://tracker.ceph.com/issues/39393

### CODE REVIEW

File: src/cls/rgw/cls_rgw_client.cc [hunk 1]
```diff
@@ -209,12 +209,13 @@ void cls_rgw_bucket_complete_op(ObjectWriteOperation& o, RGWModifyOp op, string&
   o.exec(RGW_CLASS, RGW_BUCKET_COMPLETE_OP, in);
 }
 
-static bool issue_bucket_list_op(librados::IoCtx& io_ctx, const string& oid,
-				 const cls_rgw_obj_key& start_obj,
-				 const string& filter_prefix,
-				 uint32_t num_entries, bool list_versions,
-				 BucketIndexAioManager *manager,
-				 rgw_cls_list_ret *pdata) {
+void cls_rgw_bucket_list(librados::ObjectReadOperation& op,
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_client.cc]: this just sets up the op, yes?

Comment (mattbenjamin) [reply src/cls/rgw/cls_rgw_client.cc]: that's probably a lexical pattern, but maybe in future we could make that visible in the name.

Comment (cbodley) [reply src/cls/rgw/cls_rgw_client.cc]: how about `cls_rgw_bucket_list_op()`?

Comment (mattbenjamin) [reply src/cls/rgw/cls_rgw_client.cc]: that's a good ctor, yeah

**Comment by cbodley** [comment]
updated to rename cls_rgw_bucket_list() to cls_rgw_bucket_list_op()

**Comment by cbodley** [comment]
pass qa in http://pulpito.ceph.com/cbodley-2019-04-22_17:44:29-rgw-wip-cbodley-testing-distro-basic-smithi/

many valgrind issues from ceph-mon: http://tracker.ceph.com/issues/38827
three-zone multisite failure in test_versioned_object_incremental_sync: http://tracker.ceph.com/issues/39142

**Pull request merged**